File: src/main/java/io/icednut/algorithm/exercise/graph/leet733/Solution.java
package io.icednut.algorithm.exercise.graph.leet733;

import java.util.HashSet;

class Solution {
    public int[][] floodFill(int[][] image, int sr, int sc, int color) {
        var pastColor = image[sr][sc];
        var seen = new HashSet<String>();

        dfs(sr, sc, image, seen, pastColor, color);
        return image;
    }

    private int[][] init(int[][] image) {
        var result = new int[image.length][];

        for (var i = 0; i < image.length; i++) {
            result[i] = new int[image[i].length];
        }
        return result;
    }

    private void dfs(int sr, int sc, int[][] image, HashSet<String> seen, int pastColor, int color) {
        if(sr >= image.length || sr < 0) {
            return;
        }
        if (sc >= image[sr].length || sc < 0) {
            return;
        }

        if (seen.contains("%d,%d".formatted(sr, sc))) {
            return;
        }

        seen.add("%d,%d".formatted(sr, sc));

        if (image[sr][sc] == pastColor) {
            image[sr][sc] = color;
            dfs(sr - 1, sc, image, seen, pastColor, color);
            dfs(sr + 1, sc, image, seen, pastColor, color);
            dfs(sr, sc - 1, image, seen, pastColor, color);
            dfs(sr, sc + 1, image, seen, pastColor, color);
        }
    }
}

File: src/main/kotlin/io/icednut/algorithm/exercise/graph/leet733/Solution.kt
package io.icednut.algorithm.exercise.graph.leet733

import java.util.*

class Solution {
    fun floodFill(image: Array<IntArray>, sr: Int, sc: Int, color: Int): Array<IntArray> {
        val stack = Stack<String>()
        val seen = HashSet<String>()
        val result = image.map {
            it.toMutableList()
        }
        val beforeColor = image[sr][sc]

        stack.push("${sr},${sc}")

        while (stack.isNotEmpty()) {
            val nextPos = stack.pop().split(",")
            val x = nextPos[0].toInt()
            val y = nextPos[1].toInt()

            if (result.size <= x || x < 0) {
                continue
            }

            if (result[x].size <= y || y < 0) {
                continue
            }

            val targetColor = result[x][y]

            if (!seen.contains("${x},${y}") && targetColor == beforeColor) {
                seen.add("${x},${y}")

                if (targetColor != color) {
                    result[x][y] = color
                }
                stack.push("${x - 1},${y}")
                stack.push("${x + 1},${y}")
                stack.push("${x},${y - 1}")
                stack.push("${x},${y + 1}")
            }
        }

        return result.map {
            it.toIntArray()
        }.toTypedArray()
    }
}